app/src/pages/inside/dashboardItemPage/widgetsGrid/widgetsGrid.jsx

```diff
@@ -105,56 +99,75 @@ export class WidgetsGrid extends PureComponent {
   }
 
   fetchWidgets() {
+    const { userInfo, project } = this.props;
+    const projectRole =
+      userInfo.assigned_projects[project] && userInfo.assigned_projects[project].projectRole;
+
     this.setState({ isFetching: true });
 
-    return fetch(this.props.url).then(({ widgets }) => {
-      this.setState({
-        widgets,
-        isFetching: false,
+    return fetch(this.props.url)
+      .then((data) => {
+        const isOwner = data.owner === userInfo.userId;
+
+        this.setState({
+          widgets: data.widgets,
+          isFetching: false,
+          isModifiable: canResizeAndDragWidgets(userInfo.userRole, projectRole, isOwner),
+        });
+      })
+      .catch(({ response }) => {
+        if (response.status === 404) {
+          this.props.history.push(`/#${project}/dashboard`)
+        }
       });
-    });
   }
 
   render() {
     const { widgets } = this.state;
     let Items = null;
     let height = 0; // we need to set large height to avoid double scroll
+
     if (widgets.length) {
       Items = widgets.map(({ widgetPosition: [x, y], widgetSize: [w, h], widgetId }) => {
         height += h * (rowHeight + 20);
         return (
           <div
             key={widgetId}
             className={cx('widget-view')}
-            data-grid={{ x, y, w, h, minW: 4, minH: 4 }}
+            data-grid={{ x, y, w, h, minW: 4, minH: 4, i: widgetId }}
           >
-            <Widget switchDraggable={this.switchDraggable} />
+            <Widget isModifiable={this.state.isModifiable} />
           </div>
         );
       });
     }
 
     return (
-      <div>
+      <div className={this.state.isMobile ? 'mobile ' : ''}>
         {this.state.isFetching && <SpinningPreloader />}
```

Comment:
Please add some vertical margins for preloader container(add it).  It will looks a bit prettier.

Reply:
done